### PR TITLE
Enable strict null checks

### PR DESCRIPTION
**Pull request by brcrista**
Consuming task-lib in strict mode is problematic because the library's .d.ts files don't specify what can be null.  I discussed this with @stephenmichaelf and @damccorm and we decided to give strict mode a shot.

Turning on strict null checks exposed a few places in the library that could have thrown null references under certain conditions.

It would be best to ship this with a new major version of the library.  If a consumer is compiled with strict checks and made wrong assumptions about which API functions can return `null`, this change would break their TypeScript compilation.  Also, in a few places I changed `null` to `undefined` (in order to say `string | undefined` instead of `string | null | undefined`).

I didn't turn on strict null checking in the test code.

### CODE REVIEW

File: node/toolrunner.ts [hunk 1]
```diff
@@ -581,22 +581,22 @@ export class ToolRunner extends events.EventEmitter {
         });
 
         let success = true;
-        options = this._cloneExecOptions(options);
+        const clonedOptions = this._cloneExecOptions(options);
```
Comment (brcrista) [inline node/toolrunner.ts]: `_cloneExecOptions` takes a nullable `options` and returns a non-null object.

**Comment by damccorm** [comment]
@brcrista are you still interested in reviving this like we talked about a couple weeks ago? We're now using TypeScript 3 FYI

**Comment by brcrista** [comment]
@damccorm I've updated the PR.  I'll find a time for us to go through the changes this week.

File: node/toolrunner.ts [hunk 2]
```diff
@@ -788,26 +788,22 @@ export class ToolRunner extends events.EventEmitter {
...
         });
 
-        let cp = child.spawn(this._getSpawnFileName(), this._getSpawnArgs(options), this._getSpawnOptions(options));
+        let cp = child.spawn(this._getSpawnFileName(), this._getSpawnArgs(optionsNonNull), this._getSpawnOptions(options));
 
         var stdbuffer: string = '';
         cp.stdout.on('data', (data: Buffer) => {
             this.emit('stdout', data);
 
-            if (!options.silent) {
```
Comment (damccorm) [inline node/toolrunner.ts]: Why did you get rid of this?

File: node/mock-task.ts [hunk 3]
```diff
@@ -460,13 +456,13 @@ exports.ToolRunner = trm.ToolRunner;
...
-export function getHttpProxyConfiguration(requestUrl?: string): task.ProxyConfiguration {
+export function getHttpProxyConfiguration(requestUrl?: string): task.ProxyConfiguration | null {
     return null;
 }
 
 //-----------------------------------------------------
 // Http Certificate Helper
 //-----------------------------------------------------
-export function getHttpCertConfiguration(): task.CertConfiguration {
+export function getHttpCertConfiguration(): task.ProxyConfiguration | null {
```
[the remark this answers is not in the document]
Comment (brcrista) [reply node/mock-task.ts]: oops -- copy-paste error from the other method

File: node/mock-task.ts [hunk 4]
```diff
@@ -108,52 +107,52 @@ module.exports.getSecureFileTicket = task.getSecureFileTicket;
...
-        this.m_isSocket = mockResponses['isSocket'];
+    private m_isFile: boolean = false;
+    private m_isDirectory: boolean = false;
+    private m_isBlockDevice: boolean = false;
+    private m_isCharacterDevice: boolean = false;
+    private m_isSymbolicLink: boolean = false;
+    private m_isFIFO: boolean = false;
+    private m_isSocket: boolean = false;
+
+    dev: number = 0;
```
[the remark this answers is not in the document]
Comment (brcrista) [reply node/mock-task.ts]: That looks like just an example to me.  Or did you mean something else?

File: node/mock-test.ts [hunk 5]
```diff
@@ -36,14 +34,14 @@ export class MockTestRunner {
...
+    }
 
     public stdOutContained(message: string): boolean {
-        return this.stdout && this.stdout.indexOf(message) > 0;
+        return this.stdout.indexOf(message.trim()) > 0;
     }
 
     public stdErrContained(message: string): boolean {
-        return this.stderr && this.stderr.indexOf(message) > 0;
+        return this.stderr.indexOf(message.trim()) > 0;
```
[the remark this answers is not in the document]
Comment (brcrista) [reply node/mock-test.ts]: Sounds good to me.  I can remove all of these

[the remark this answers is not in the document]
Comment (brcrista) [reply node/mock-test.ts]: Actually, I see the incoming messages get trimmed too.  I wonder if there's something more subtle here about how ##vso commands work.  I'll just revert this change for now.

File: node/toolrunner.ts (hunk 2 above)
Comment (brcrista) [reply node/toolrunner.ts]: Bad merge -- I thought this had moved when I was resolving conflicts.

File: node/mock-task.ts (hunk 4 above)
[the remark this answers is not in the document]
Comment (damccorm) [reply node/mock-task.ts]: Oh yeah, as I look further you're right. Looks like they don't actually have defaults, so I'm comfortable with this.

**Comment by damccorm** [comment]
Looks like the failing check is just flakiness, requeued a successful build here - https://dev.azure.com/mseng/PipelineTools/_build/results?buildId=8812994

Overall LGTM

**Pull request merged**